src/Shared/FileMatcher.cs

```diff
@@ -1094,215 +1119,281 @@ private static string RegularExpressionFromFileSpec
-                matchFileExpression.Replace("<:dirseparator:>.<:dirseparator:>", "<:dirseparator:>");
-                matchFileExpression.Replace("<:dirseparator:><:dirseparator:>", "<:dirseparator:>");
-                matchFileExpression.Replace("<:fixeddir:>.<:dirseparator:>.<:dirseparator:>", "<:fixeddir:>.<:dirseparator:>");
-                matchFileExpression.Replace("<:dirseparator:>.<:endfilename:>", "<:endfilename:>");
-                matchFileExpression.Replace("<:filename:>.<:endfilename:>", "<:filename:><:endfilename:>");
+        /// <summary>
+        /// Append the regex equivalents for character sequences in the wildcard directory part of a filespec:
+        ///
+        /// (1) The leading **\ if existing
+        ///
+        /// (2) Each occurence of recursive wildcard \**\
+        /// 
+        /// (3) Common filespec characters
+        /// </summary>
+        private static void AppendRegularExpressionFromWildcardDirectory(ReuseableStringBuilder regex, string wildcardDir)
+        {
+            regex.Append(FileSpecRegexParts.FixedDirWildcardDirSeparator);
 
-                ErrorUtilities.VerifyThrow(matchFileExpression.Length <= sizeBefore,
-                    "Expression reductions cannot increase the length of the expression.");
-            } while (matchFileExpression.Length < sizeBefore);
+            bool hasRecursiveOperatorAtStart = wildcardDir.Length > 2 && wildcardDir[0] == '*' && wildcardDir[1] == '*';
 
-            /*
-             * Collapse **\** into **.
-             */
-            do
+            if (hasRecursiveOperatorAtStart)
             {
-                sizeBefore = matchFileExpression.Length;
-                matchFileExpression.Replace(recursiveDirectoryMatch + "<:dirseparator:>" + recursiveDirectoryMatch, recursiveDirectoryMatch);
+                regex.Append(FileSpecRegexParts.LeftDirs);
+            }
+            int startIndex = hasRecursiveOperatorAtStart ? IndexOfNextNonCollapsibleChar(wildcardDir, 3) : 0;
 
-                ErrorUtilities.VerifyThrow(matchFileExpression.Length <= sizeBefore,
-                    "Expression reductions cannot increase the length of the expression.");
-            } while (matchFileExpression.Length < sizeBefore);
+            for (int i = startIndex; i < wildcardDir.Length; i = IndexOfNextNonCollapsibleChar(wildcardDir, i + 1))
+            {
+                char ch = wildcardDir[i];
+                bool isRecursiveOperator = i < wildcardDir.Length - 2 && wildcardDir[i + 1] == '*' && wildcardDir[i + 2] == '*';
 
-            /*
-             * Call out legal recursion operators:
-             * 
-             *        fixed-directory + **\
-             *        \**\
-             *        **\**
-             * 
-             */
-            do
+                if (isRecursiveOperator)
+                {
+                    regex.Append(FileSpecRegexParts.MiddleDirs);
+                    i += 3;
```

Comment:
Asserts that the recursive chars are surrounded by file separators would help with documentation.